File: app/src/main/java/it/gruppoinfor/home2work/adapters/CompanySpinnerAdapter.java
package it.gruppoinfor.home2work.adapters;

import android.app.Activity;
import android.content.Context;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ArrayAdapter;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import it.gruppoinfor.home2work.R;
import it.gruppoinfor.home2workapi.model.Company;


public class CompanySpinnerAdapter extends ArrayAdapter<Company> {

    private Context context;
    private ArrayList<Company> companies;
    private LayoutInflater inflater;

    public CompanySpinnerAdapter(Activity context, List<Company> companies) {
        super(context, R.layout.item_company_spinner, R.id.company_name, companies);
        this.context = context;
        this.companies = new ArrayList<>(companies);
        inflater = context.getLayoutInflater();
    }

    @Override
    public View getDropDownView(int position, @Nullable View view, @NonNull ViewGroup parent) {
        ViewHolder holder;
        Company company = companies.get(position);

        if (view == null) {
            view = inflater.inflate(R.layout.item_company_spinner, parent, false);
            holder = new ViewHolder(view);
            view.setTag(holder);
        } else {
            holder = (ViewHolder) view.getTag();
        }

        holder.companyName.setText(company.getName());
        holder.companyAddress.setText(company.getAddress().getCity());

        return view;
    }

    @NonNull
    @Override
    public View getView(int position, @Nullable View view, @NonNull ViewGroup parent) {
        ViewHolder holder;
        Company company = companies.get(position);

        if (view == null) {
            view = inflater.inflate(R.layout.item_company_spinner, parent, false);
            holder = new ViewHolder(view);
            view.setTag(holder);
        } else {
            holder = (ViewHolder) view.getTag();
        }

        holder.companyName.setText(company.getName());
        holder.companyAddress.setText(company.getAddress().getCity());

        return view;
    }

    @Override
    public int getCount() {
        return companies.size() - 1;
    }

    static class ViewHolder {
        @BindView(R.id.company_name)
        TextView companyName;
        @BindView(R.id.company_address)
        TextView companyAddress;

        ViewHolder(View view) {
            ButterKnife.bind(this, view);
        }
    }
}

File: App/src/main/java/it/gruppoinfor/home2work/adapters/CompanySpinnerAdapter.kt
package it.gruppoinfor.home2work.adapters

import android.app.Activity
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ArrayAdapter
import android.widget.TextView
import it.gruppoinfor.home2work.R
import it.gruppoinfor.home2workapi.model.Company
import kotlinx.android.synthetic.main.item_company_spinner.view.*
import java.util.*


@Suppress("NAME_SHADOWING")
class CompanySpinnerAdapter(activity: Activity, companies: List<Company>) : ArrayAdapter<Company>(activity, R.layout.item_company_spinner, R.id.company_name, companies) {

    private val mCompanies: ArrayList<Company>
    private val mInflater: LayoutInflater

    init {
        mCompanies = ArrayList(companies)
        mInflater = activity.layoutInflater
    }

    override fun getDropDownView(position: Int, view: View?, parent: ViewGroup): View {
        var view = view
        val holder: ViewHolder
        val company = mCompanies[position]

        if (view == null) {
            view = mInflater.inflate(R.layout.item_company_spinner, parent, false)
            holder = ViewHolder(view)
            view!!.tag = holder
        } else {
            holder = view.tag as ViewHolder
        }

        holder.companyName.text = company.name
        holder.companyAddress.text = company.address.city

        return view
    }

    override fun getView(position: Int, view: View?, parent: ViewGroup): View {
        var view = view
        val holder: ViewHolder
        val company = mCompanies[position]

        if (view == null) {
            view = mInflater.inflate(R.layout.item_company_spinner, parent, false)
            holder = ViewHolder(view)
            view!!.tag = holder
        } else {
            holder = view.tag as ViewHolder
        }

        holder.companyName.text = company.name
        holder.companyAddress.text = company.address.city

        return view
    }

    override fun getCount(): Int {
        return mCompanies.size - 1
    }

    class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val companyName: TextView = itemView.company_name
        val companyAddress: TextView = itemView.company_address
    }
}
